examples/step-86/step-86.cc

```diff
@@ -0,0 +1,1387 @@
+    convergence_table.evaluate_convergence_rates(
+      "LevelSet dofs", ConvergenceTable::reduction_rate_log2);
+
+    convergence_table.add_value("Active dofs", dof_handler_size);
+    convergence_table.evaluate_convergence_rates(
+      "Active dofs", ConvergenceTable::reduction_rate_log2);
+
+    convergence_table.add_value("CPU time", time);
+    convergence_table.evaluate_convergence_rates(
+      "CPU time", ConvergenceTable::reduction_rate_log2);
+
+    convergence_table.add_value("Iterations", iterations);
+
+    convergence_table.add_value("L2 Error", error_L2);
+    convergence_table.evaluate_convergence_rates(
+      "L2 Error", ConvergenceTable::reduction_rate_log2);
+    convergence_table.set_scientific("L2 Error", true);
+
+    convergence_table.add_value("H1 error", error_semiH1);
+    convergence_table.evaluate_convergence_rates(
+      "H1 error", ConvergenceTable::reduction_rate_log2);
+    convergence_table.set_scientific("H1 error", true);
+
+    convergence_table.add_value("Stab norm", error_stab);
+    convergence_table.evaluate_convergence_rates(
+      "Stab norm", ConvergenceTable::reduction_rate_log2);
+    convergence_table.set_scientific("Stab norm", true);
+
+    pcout << std::endl;
+#ifdef USE_TRILINOS
+    if (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0)
+      convergence_table.write_text(pcout.get_stream());
+#else
+    convergence_table.write_text(std::cout);
+#endif
+    pcout << std::endl;
+  }
+
+  // The method performs VTK output of preliminary mesh refinements of geometry
+  // approximation.
+  template <int dim>
+  void LaplaceBeltramiSolver<dim>::output_levelset(unsigned int n) const
+  {
+    pcout << "Writing vtu file for surface... ";
+    Timer        timer;
+    DataOut<dim> data_out;
+    data_out.add_data_vector(level_set_dof_handler, level_set, "level_set");
+    data_out.add_data_vector(activeFE_indicator, "ref_indicator");
+#ifdef USE_TRILINOS
+    Vector<float> subdomain(triangulation.n_active_cells());
+    for (unsigned int i = 0; i < subdomain.size(); ++i)
+      subdomain(i) = triangulation.locally_owned_subdomain();
+    data_out.add_data_vector(subdomain, "subdomain");
+
+    data_out.build_patches();
+    // have to create the directory output
+    data_out.write_vtu_in_parallel("step-86-surface_" + std::to_string(n) +
+                                     ".vtu",
+                                   MPI_COMM_WORLD);
+//
+//    data_out.write_vtu_with_pvtu_record(
+//      "./", "step86-surface", n, MPI_COMM_WORLD, 2, 8);
+#else
+    data_out.build_patches();
+    std::ofstream output("step86-surface_" + std::to_string(n) + ".vtu");
+    data_out.write_vtu(output);
+#endif
+    timer.stop();
+    pcout << "took (" << timer.cpu_time() << "s)" << std::endl;
+  }
+
+  // The method performs VTK output of the TraceFEM solution.
+  template <int dim>
+  void LaplaceBeltramiSolver<dim>::output_final() const
+  {
+    pcout << "Writing vtu file... ";
```

Comment:
lots of duplication here. Maybe the difference can be controlled by a flag to a single method?